### ISSUE TITLE
ARM64 FFI zeroes out varargs

### ISSUE THREAD

**Issue by rweichler** [LuaJIT/LuaJIT#141]
I realize that it is still in beta. If this is due to the feature not being implemented, or something like that, then I'm sorry for wasting an issue on this.

Here's an example:

Lua script:
```lua
local ffi = require 'ffi'
ffi.load("add_em_up", true)
ffi.cdef [[
void add_em_up(int count, ...);
]]

ffi.C.add_em_up(2, ffi.cast("int", 1), ffi.cast("int", 1))
```

c library:
```c
#include <stdarg.h>
#include <stdio.h>

void add_em_up (int count, ...)
{
  va_list ap;
  int i;

  va_start (ap, count);

  for (i = 0; i < count; i++) {
    int x = va_arg(ap, int);
    printf("arg[%d] = %d\n", i, x);
  }

  va_end (ap);
}
```

Output:
```
arg[0] = 0
arg[1] = 0
```

**Comment by bigbes** [comment]
You must use `double` type instead of `int`. It's pointed in [documentation](http://luajit.org/ext_ffi_semantics.html#convert_fromlua).
Here's output if you'll replace `int` with `double`:
```
arg[0] = 1.000000
arg[1] = 1.000000
```

**Comment by rweichler** [comment]
@bigbes That actually is not the issue I was trying to outline. To humor you, changing it to `double` like you said produces this output on arm64:

```
arg[0] = 0.000000
arg[1] = 0.000000
```

However, I admit I gave a bad example. I quickly wrote that up after discovering that passing strings resolves to `NULL`. I should have just pasted that code instead of needlessly complicating things and causing a digression.

**Comment by MikePall** [comment]
You didn't specify which OS and toolchain you use. Your example works fine on Linux/ARM64.

Maybe you're using iOS/ARM64? I can't test that myself, but the code follows the description of the required ABI changes from Apple.

The relevant code is [here](https://github.com/LuaJIT/LuaJIT/blob/64c6da6b21e5888e7fac0a9a92dbde313258bfb1/src/lj_ccall.c#L337) Note that `isva` is set to 1 for all arguments past the fixed arguments. And the default for `CCALL_HANDLE_REGARG` is to fall through to stack allocation of an argument.

**Comment by rweichler** [comment]
@MikePall Sorry. Yeah, iOS/ARM64, jailbroken. And thanks. I'll take a look at the code later and see if I can fix it myself.

**Comment by MikePall** [comment]
Thank you for the offer, but I _can_ test ARM64 code (with QEMU). This is more about access to iOS dev-tools and their NDA, which I'm very, very careful to avoid.

It would help to get the assembler output of your C snippet (`-O2 -S`), to glean how the platform compiler actually handles vararg access. Maybe I've misinterpreted the (slightly confusing) description of the [iOS ABI divergences](https://developer.apple.com/library/ios/documentation/Xcode/Conceptual/iPhoneOSABIReference/Articles/ARM64FunctionCallingConventions.html).

**Comment by corsix** [comment]
If it helps:
```
[littlegate /Users/corsix]$ cat ios.c
int printf(const char*, ...);
void add_em_up(int count, ...);
int main() {
  add_em_up(2, 1, 1);
  return printf("%d %s %f", 11, "22", 33.3);
}

#include <stdarg.h>

void add_em_up (int count, ...)
{
  va_list ap;
  int i;

  va_start (ap, count);

  for (i = 0; i < count; i++) {
    int x = va_arg(ap, int);
    printf("arg[%d] = %d\n", i, x);
  }

  va_end (ap);
}
[littlegate /Users/corsix]$ clang -arch arm64 -mios-version-min=6.1 -O2 -S ios.c
[littlegate /Users/corsix]$ cat ios.s
	.section	__TEXT,__text,regular,pure_instructions
	.ios_version_min 6, 1
	.globl	_main
	.align	2
_main:                                  ; @main
; BB#0:
	stp	x29, x30, [sp, #-16]!
	mov	 x29, sp
	sub	sp, sp, #32             ; =32
	orr	w8, wzr, #0x1
	stp	 x8, x8, [sp]
	orr	w0, wzr, #0x2
	bl	_add_em_up
	orr	x8, xzr, #0x6666666666666666
	movk	x8, #0xa666, lsl #32
	movk	x8, #0x4040, lsl #48
	str	x8, [sp, #16]
Lloh0:
	adrp	x8, L_.str1@PAGE
Lloh1:
	add	x8, x8, L_.str1@PAGEOFF
	str	x8, [sp, #8]
	movz	w8, #0xb
	str	 x8, [sp]
Lloh2:
	adrp	x0, L_.str@PAGE
Lloh3:
	add	x0, x0, L_.str@PAGEOFF
	bl	_printf
	mov	 sp, x29
	ldp	x29, x30, [sp], #16
	ret
	.loh AdrpAdd	Lloh0, Lloh1
	.loh AdrpAdd	Lloh2, Lloh3

	.globl	_add_em_up
	.align	2
_add_em_up:                             ; @add_em_up
; BB#0:
	stp	x22, x21, [sp, #-48]!
	stp	x20, x19, [sp, #16]
	stp	x29, x30, [sp, #32]
	add	x29, sp, #32            ; =32
	sub	sp, sp, #32             ; =32
	mov	 x19, x0
	add	x8, x29, #16            ; =16
	str	x8, [sp, #24]
	cmp	 w19, #1                ; =1
	b.lt	LBB1_3
; BB#1:
	movz	w21, #0
Lloh4:
	adrp	x20, L_.str2@PAGE
Lloh5:
	add	x20, x20, L_.str2@PAGEOFF
LBB1_2:                                 ; =>This Inner Loop Header: Depth=1
	ldr	x8, [sp, #24]
	add	x9, x8, #8              ; =8
	str	x9, [sp, #24]
	ldr	 w8, [x8]
	stp	 x21, x8, [sp]
	mov	 x0, x20
	bl	_printf
	add	w21, w21, #1            ; =1
	cmp	 w19, w21
	b.ne	LBB1_2
LBB1_3:                                 ; %._crit_edge
	sub	sp, x29, #32            ; =32
	ldp	x29, x30, [sp, #32]
	ldp	x20, x19, [sp, #16]
	ldp	x22, x21, [sp], #48
	ret
	.loh AdrpAdd	Lloh4, Lloh5

	.section	__TEXT,__cstring,cstring_literals
L_.str:                                 ; @.str
	.asciz	"%d %s %f"

L_.str1:                                ; @.str1
	.asciz	"22"

L_.str2:                                ; @.str2
	.asciz	"arg[%d] = %d\n"


.subsections_via_symbols
```

**Comment by corsix** [comment]
@rweichler: As a diagnostic aid, have you tried passing say ten varargs rather than just two?

As you're getting zeroes rather than the expected values, it initially feels like LuaJIT is passing the varargs on the stack (as per the usual iOS ARM64 ABI), but your C library is trying to read varargs from registers (which LuaJIT helpfully sets to zero to get some determinism in case of misdeclaration).

**Comment by MikePall** [comment]
The assembler output clearly shows it's fetching the varargs from the first stack slots. And LuaJIT's FFI puts them at that exact spot for an iOS/ARM64 build. The assembler code matches my understanding of the ABI and it should just work all together.

OK, so I tried to emulate iOS and forced the ABI in `lj_ccall.c`. I've checked with QEMU/gdb that the arguments actually end up on the stack in the right position.

Well, I'm stumped on this one.

The only thing I could imagine is something went wrong in your LuaJIT build (`LJ_TARGET_IOS` not being set) or your C library build (what @corsix is suggesting).

Or maybe it's calling the wrong function? Have you tried:

```lua
local lib = ffi.load("add_em_up")
...
lib.add_em_up(...)
```
